packages/fetchai/connections/soef/connection.py

```diff
@@ -91,106 +147,349 @@ def __init__(
+        value = service_description.values.get("value", NOT_SPECIFIED)
+
+        if key is None or value is NOT_SPECIFIED:  # pragma: nocover
+            raise SOEFException.error("Bad values provided!")
+
+        await self._set_service_key(key, value)
+
+    async def _generic_oef_command(
+        self, command, params=None, unique_page_address=None, check_success=True
+    ):
+        """
+        Set service key from service description.
+
+        :param service_description: Service description
+        :return None
+        """
+        params = params or {}
+        logger.debug(f"Perform `{command}` with {params}")
+        url = parse.urljoin(
+            self.base_url, unique_page_address or self.unique_page_address
+        )
+        response_text = await self._request_text(
+            "get", url=url, params={"command": command, **params}
         )
-        return is_compatible
+        try:
+            root = ET.fromstring(response_text)
+            assert root.tag == "response"
+            if check_success:
+                el = root.find("./success")
+                assert el is not None, "No success element"
+                assert str(el.text).strip() == "1", "Success is not 1"
+            logger.debug(f"`{command}` SUCCSESS!")
+            return response_text
+        except Exception as e:
+            raise SOEFException.error(f"`{command}` error: {response_text}: {[e]}")
 
-    def _register_agent(self) -> None:
+    async def _set_service_key(self, key: str, value: Union[str, int, float]) -> None:
+        """
+        Perform set service key command.
+
+        :param key: key to set
+        :param value: value to set
+        :return None:
+        """
+        await self._generic_oef_command("set_service_key", {"key": key, "value": value})
+
+    async def _remove_service_key_handler(
+        self, service_description: Description
+    ) -> None:
+        """
+        Remove service key from service description.
+
+        :param service_description: Service description
+        :return None
+        """
+        self._sure_data_model(service_description, "remove_service_key")
+        key = service_description.values.get("key", None)
+
+        if key is None:  # pragma: nocover
+            raise SOEFException.error("Bad values provided!")
+
+        await self._remove_service_key(key)
+
+    async def _remove_service_key(self, key: str) -> None:
+        """
+        Perform remove service key command.
+
+        :param key: key to remove
+        :return None:
+        """
+        await self._generic_oef_command("remove_service_key", {"key": key})
+
+    async def _register_location_handler(
+        self, service_description: Description
+    ) -> None:
+        """
+        Register service with location.
+
+        :param service_description: Service description
+        :return None
+        """
+        self._sure_data_model(service_description, "location_agent")
```

Comment:
magic values like `location_agent` and `remove_service_key` should be specified as reusable constants!

Reply:
done